pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,844 @@
+	)
+
+	metricTimeToApplySyncSetResource = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "hive_syncsetinstance_apply_duration_seconds",
+			Help:    "Time to apply individual resources in a syncset, labeled by type of apply and result.",
+			Buckets: []float64{0.5, 1, 3, 5, 10, 20, 30},
+		},
+		[]string{"type", "result"},
+	)
+)
+
+func init() {
+	metrics.Registry.MustRegister(metricTimeToApplySyncSet)
+	metrics.Registry.MustRegister(metricTimeToApplySelectorSyncSet)
+	metrics.Registry.MustRegister(metricResourcesApplied)
+	metrics.Registry.MustRegister(metricTimeToApplySyncSetResource)
+}
+
+// Add creates a new clustersync Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	r, err := NewReconciler(mgr)
+	if err != nil {
+		return err
+	}
+	return AddToManager(mgr, r)
+}
+
+// NewReconciler returns a new ReconcileClusterSync
+func NewReconciler(mgr manager.Manager) (*ReconcileClusterSync, error) {
+	logger := log.WithField("controller", ControllerName)
+	reapplyInterval := defaultReapplyInterval
+	if envReapplyInterval := os.Getenv(reapplyIntervalEnvKey); len(envReapplyInterval) > 0 {
+		var err error
+		reapplyInterval, err = time.ParseDuration(envReapplyInterval)
+		if err != nil {
+			log.WithError(err).WithField("reapplyInterval", envReapplyInterval).Errorf("unable to parse %s", reapplyIntervalEnvKey)
+			return nil, err
+		}
+	}
+	log.WithField("reapplyInterval", reapplyInterval).Info("Reapply interval set")
+	c := controllerutils.NewClientWithMetricsOrDie(mgr, ControllerName)
+	return &ReconcileClusterSync{
+		Client:                c,
+		logger:                logger,
+		reapplyInterval:       reapplyInterval,
+		resourceHelperBuilder: resourceHelperBuilderFunc,
+		remoteClusterAPIClientBuilder: func(cd *hivev1.ClusterDeployment) remoteclient.Builder {
+			return remoteclient.NewBuilder(c, cd, ControllerName)
+		},
+	}, nil
+}
+
+func resourceHelperBuilderFunc(restConfig *rest.Config, logger log.FieldLogger) (resource.Helper, error) {
+	return resource.NewHelperFromRESTConfig(restConfig, logger), nil
+}
+
+// AddToManager adds a new Controller to mgr with r as the reconcile.Reconciler
+func AddToManager(mgr manager.Manager, r *ReconcileClusterSync) error {
+	// Create a new controller
+	c, err := controller.New("clusterSync-controller", mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: controllerutils.GetConcurrentReconciles()})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ClusterDeployment
+	if err := c.Watch(&source.Kind{Type: &hivev1.ClusterDeployment{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	// Watch for changes to SyncSets
+	if err := c.Watch(
+		&source.Kind{Type: &hivev1.SyncSet{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(requestsForSyncSet),
```

Comment:
It _could_. But the signature of `requestForSyncSet` already matches the type needed. Having `requestForSyncSet` return a function in order to avoid a cast is overkill.

Reply:
Ack. The two `ToRequests` don't seem different to me; I was just looking for consistency.